src/com/cdd/bao/template/SchemaTree.java

```diff
@@ -138,7 +142,83 @@ public String toString()
 		}
 		return ancestors.toArray(new String[ancestors.size()]);
 	}
+
+	// candidates is a list of (parentURI, Node) pairs, where each node is candidate
+	// for this SchemaTree, and should preset label, description, and uri fields
+	// return true if SchemaTree was changed, false otherwise
+	public boolean addNodes(List<Pair<String, Node>> candidates)
```

Comment:
This is what I was suggesting earlier today:

```
List<Pair<String, Node>> candidates = <shallow clone of nodes>;
while (candidates.size() > 0)
{
	boolean anything = false;
	for (Iterator<Pair<String, Node>> it = candidates.iterator(); it.hasNext();)
	{
		if <parent is in the tree>
		{
			<create and add the node...>
			anything = true;
			it.remove();
		}
	}
	if (!anything) break;
}
```

It's shorter and cleaner. Then you could follow that up with

`if (candidates.size() == nodes.size()) return; // nothing happened`

And also careful with the `flattenTree` function, it doesn't currently clear out the existing list.

Reply:
I like this.  However, I think we need logic that coerces provisionals whose parents are provisionals to go at the end of the candidates list.   I guess we can incorporate your logic AND this coercive condition and it will probably be faster than what I have.  I'll do that shortly.

Reply:
Looking at `SchemaTree.addNodes` more closely, it's clear that the *interesting* scenario is when a provisional term has a parent amongst the candidate provisional terms.  We need to treat that condition properly; hence, my implementation may be less concise, but properly addresses the provisional-to-provisional-parent scenario.